pkg/plan/track_response_builder.go

```diff
@@ -0,0 +1,216 @@
+)
+
+// buildTrackResponse takes a models.DeploymentResources and iterates over the
+// deployment's resources to obtain a TrackResponse from either the Pending,
+// Current or History fields from the resource's plan. It returns []TrackResponse
+// each of them being an individual update about a resource's plan. When
+// getCurrentPlan is set to true, the plan which is looked up is either the current
+// or the last plan in the resource's plan history in the case that the resource
+// does not have any current plan, which is a common case for resources which failed
+// to create properly.
+func buildTrackResponse(res *models.DeploymentResources, getCurrentPlan bool) []TrackResponse {
+	var pending = make([]TrackResponse, 0)
+	for _, info := range res.Elasticsearch {
+		p, err := parseResourceInfo(info, "elasticsearch", getCurrentPlan)
+		if err != nil {
+			continue
+		}
+		pending = append(pending, p)
+	}
+
+	for _, info := range res.Kibana {
+		p, err := parseResourceInfo(info, "kibana", getCurrentPlan)
+		if err != nil {
+			continue
+		}
+		pending = append(pending, p)
+	}
+
+	for _, info := range res.Apm {
+		p, err := parseResourceInfo(info, "apm", getCurrentPlan)
+		if err != nil {
+			continue
+		}
+		pending = append(pending, p)
+	}
+
+	for _, info := range res.Appsearch {
+		p, err := parseResourceInfo(info, "appsearch", getCurrentPlan)
+		if err != nil {
+			continue
+		}
+		pending = append(pending, p)
+	}
+
+	return pending
+}
+
+// parseResourceInfo takes in a <kind>ResourceInfo type along with the Kind to
+// be able to obtain the resource's plan using reflection, which is deferred to
+// getPlanStepInfo. This function builds the TrackResponse structure.
+func parseResourceInfo(info interface{}, kind string, getCurrentPlan bool) (TrackResponse, error) {
+	stepLog, err := getPlanStepInfo(info, getCurrentPlan)
+	if err != nil {
+		return TrackResponse{}, err
+	}
+
+	var id, refID string
+	if v := reflect.ValueOf(info); v.IsValid() {
+		id, refID = stringPFieldValue(v, "ID"), stringPFieldValue(v, "RefID")
+	}
+
+	step, err := GetStepName(stepLog)
+	if step == "" {
+		return TrackResponse{}, ErrPlanFinished
+	}
+
+	return TrackResponse{
+		Kind:     kind,
+		ID:       id,
+		RefID:    refID,
+		Step:     step,
+		Err:      err,
+		Finished: step == planCompleted,
+		Duration: getPlanDuration(stepLog),
+	}, nil
+}
+
+// getPlanStepInfo takes in a resource's info and returns a slice of
+// ClusterPlanStepInfo, which is obtained by accessing the named fields via
+// reflection. The flow might be confusing but in a nutshell it tries to:
+//   1. Obtain the Pending plan step log.
+//   2. Obtain the Current plan step log when getCurrentPlan is true.
+//   3. (if getCurrentPlan == true and the Current plan is empty) obtains the
+//       "Current" plan accessing the last item in the plan history slice.
+func getPlanStepInfo(wl interface{}, getCurrentPlan bool) ([]*models.ClusterPlanStepInfo, error) {
```

Comment:
I'm not sure what `wl` stands for 🤔 can we change the name of this variable?

Reply:
it stands for workload, but i'll change the receiver name